### PR TITLE
Add Argument viewhelper

### LINKED ISSUE #427: Feature: slots
The `contentAs` argument provided by Fluid is very useful to pass arbitrary content as variable to a partial/section.

However, this is limited to a single variable, if multiple pieces of content should be passed, one has to e.g. resort to doing inline `f:render` calls within `arguments` instead.

A nice solution to this would be a system similar to [Vue.js component slots](https://vuejs.org/v2/guide/components-slots.html).

### Basic example

`Example.html` template:

```html
<f:render partial="Foo" arguments="{as: 'usual'}">
Some content with<f:if condition="{lie}">out</f:if> regular Fluid.
</f:render>
```

`Foo.html` partial:

```html
Just the {as}:
<f:slot/>
```

### Named slots example

`Example.html` template:

```html
<f:render partial="BaseLayout">
  <f:template slot="header">
    <h1>Here might be a page title</h1>
  </f:template>

  <p>A paragraph for the main content.</p>
  <p>And another one.</p>

  <f:template slot="footer">
    <p>Here's some contact info</p>
  </f:template>
</f:render>
```

`BaseLayout.html` partial:

```html
<div class="container">
  <header>
    <f:slot name="header"/>
  </header>
  <main>
    <!-- Slot without name renders the content not encapsulated in `f:template` -->
    <f:slot/>
  </main>
  <footer>
    <f:slot name="footer"/>
  </footer>
</div>
```

If we stick to the same naming or use something like `argument` is up to discussion of course as is this feature in general.

### PR DESCRIPTION
**Pull request by sascha-egerer**
Add a new `f:argument` viewhelper to be able to add
arguments to a node by using `f:argument` as child node.

Usage example:
```
<my:node>
	<f:argument name="content">Hello World</f:argument>
</my:node>
```

The sequencer does now put validation errors into a stack
so we are able to revalidate a node before it is closed but
children have been parsed. So we can still show the initial
problem if e.g. a required property was not defined.

Resolves #427

### CODE REVIEW

**Comment by mbrodala** [comment]
Can you add a few tests for this?

File: src/ViewHelpers/ArgumentViewHelper.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+use TYPO3Fluid\Fluid\Component\EmbeddedComponentInterface;
+use TYPO3Fluid\Fluid\Component\SequencingComponentInterface;
+use TYPO3Fluid\Fluid\Core\Parser\Sequencer;
+use TYPO3Fluid\Fluid\Core\Rendering\RenderingContextInterface;
+use TYPO3Fluid\Fluid\Core\ViewHelper\AbstractViewHelper;
+
+/**
+ * Argument assigning ViewHelper
+ *
+ * Assigns an argument for a parent ViewHelper call when
```
Comment (NamelessCoder) [inline src/ViewHelpers/ArgumentViewHelper.php:19]: Drop "when the parent ViewHelper supports it." because every ViewHelper will support this method.

I would also reword to "Assigns an argument to the parent node" since this works for every node that supports arguments (Atom, ViewHelper).

File: src/ViewHelpers/ArgumentViewHelper.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+{
+    protected $escapeOutput = false;
+
+    /**
+     * @return void
+     */
+    public function initializeArguments()
+    {
+        parent::initializeArguments();
+        $this->registerArgument('name', 'string', 'Name of the parameter', true);
```
Comment (NamelessCoder) [inline src/ViewHelpers/ArgumentViewHelper.php:50]: Argument "value" also needs to be registered (as optional) with description saying that "If not provided, argument value is taken from tag content or inline-passed child").

File: src/Core/Parser/Sequencer.php [hunk 3]
```diff
@@ -438,7 +445,15 @@ protected function sequenceTagNode(): ?ComponentInterface
 
                     if (!$closing || $selfClosing) {
                         $viewHelperNode = $viewHelperNode ?? $this->resolver->createViewHelperInstance($namespace, (string) $method);
-                        $viewHelperNode->onOpen($this->renderingContext)->getArguments()->validate();
+                        try {
+                            $viewHelperNode->onOpen($this->renderingContext)->getArguments()->validate();
+                        } catch (Exception $e) {
+                            $argumentValidationError = $this->createErrorAtPosition(
+                                $e->getMessage(),
+                                1621337984
```
Comment (NamelessCoder) [inline src/Core/Parser/Sequencer.php:453]: Better resolve the exception code from `$e->getCode()` instead.

File: src/Core/Parser/Sequencer.php [hunk 4]
```diff
@@ -486,6 +501,36 @@ protected function sequenceTagNode(): ?ComponentInterface
...
+                    $nodeHash = spl_object_hash($viewHelperNode);
+                    // Check if we already have a validation error on the sequencing error stack
+                    foreach ($this->nodeSequencingExceptionStack[$nodeHash] ?? [] as $sequencingException) {
+                        if ($sequencingException->getCode() === 1621337984) {
+                            $argumentValidationError = $sequencingException;
+                            break;
+                        }
+                    }
+
+                    try {
```
Comment (NamelessCoder) [inline src/Core/Parser/Sequencer.php:514]: This try/catch, including the throw, should only happen if `$argumentValidationError !== null` since we don't want to re-validate arguments on every valid node - only those that previously added an error to the stack.

Then you can also skip the condition below and directly throw the error:

```php
throw ($argumentValidationError ?? $this->createErrorAtPosition(...));
```

...and better resolve the exception code from `$e->getCode()` instead of hardcoding it here.

File: src/Core/Parser/Sequencer.php [hunk 5]
```diff
@@ -191,13 +196,14 @@ public function sequence(): ComponentInterface
                     try {
                         $childNode = $this->sequenceTagNode();
                     } catch (Exception $exception) {
+                        if (!$exception instanceof SequencingException) {
```
Comment (NamelessCoder) [inline src/Core/Parser/Sequencer.php:199]: This changed block seems redundant - or am I overlooking something? If redundant, please revert this change block.

File: src/Core/Parser/Sequencer.php [hunk 6]
```diff
@@ -122,6 +122,11 @@ class Sequencer
      */
     protected $nodeStack = [];
 
+    /**
+     * @var array<string, array<SequencingException>>
```
Comment (NamelessCoder) [inline src/Core/Parser/Sequencer.php:126]: `@var SequencingException[]` is preferred.

**Review by NamelessCoder — CHANGES_REQUESTED** [review]

File: src/Core/Parser/Sequencer.php (hunk 6 above)
Comment (ohader) [reply src/Core/Parser/Sequencer.php:126]: `SequencingException[]` means `list<SequencingException>`, means `array<int, SequencingException>` - which does not reflect using `$nodeHash` as string.

**Comment by NamelessCoder** [comment]
@ohader It doesn't actually change the argument *definitions*, it merely defines an argument to pass to the atom that is rendered. These two things are equivalent:

```xml
<my:node arg="Value" />
<!-- or -->
<my:node>
    <f:argument name="arg" value="Value" />
</my:node>
```

The intention is to provide a way to pass larger argument values, such as a big block of text that would be horrible to put into a tag argument. You could also compare it to this:

```xml
<f:variable name="text">
Big, big text with "quotes" and other stuff that doesn't play well as tag attribute value.
</f:variable>

<my:node arg="{text}" />
```

Basically just a quality-of-life type of feature.

Side note; for Atoms we distinguish between a _parameter_ (which is an ArgumentDefinition) and an _argument_ (which is the value assigned to a parameter and passed to the Atom).

**Comment by ohader** [comment]
> Side note; for Atoms we distinguish between a _parameter_ (which is an ArgumentDefinition) and an _argument_ (which is the value assigned to a parameter and passed to the Atom).

I hope you do see the problem here and the confusion it creates... 😉

**Comment by NamelessCoder** [comment]
> I hope you do see the problem here and the confusion it creates... 😉

Mnjaaah, yes and no. If we didn't make this distinction then you couldn't tell an argument value from an argument definition (which then caused the confusion leading to your question today). If we do make the distinction then of course people have to be aware that there is actually a distinction. _Damned if we do and damned if we don't_ :)

BTW I'd personally recommend to make this distinction also when speaking about PHP itself, we tend to just use the name "argument" there for both the definition and the runtime value. It's just easier to talk about if there's a distinction.

**Comment by mbrodala** [comment]
@NamelessCoder I think what @ohader was hinting at is that "argument definition" / "argument" would maybe be clearer than "parameter" / "argument". At least that's how I see it and it would allow for using "parameter definition" / "parameter" interchangeably just fine.

**Comment by lolli42** [comment]
Hey.
As outlined in https://github.com/TYPO3/Fluid/issues/591, we'll close v3 related PR's for now.
I hope this is fine with you, and we may catch up on v3 codebase later again, in case development continues to happen on it.
